File: libs/go/servicekit/component.go
// Copyright 2026 Mindclade. All rights reserved.
// Confidential, proprietary, and trade-secret information.

package servicekit

import (
	"context"
	"runtime/debug"
	"strings"
	"unicode"
)

const maxNameLength = 128

// Hook is a lifecycle function. Hooks must honor context cancellation.
type Hook func(context.Context) error

// Probe checks one liveness or readiness condition. Probes must be safe for
// concurrent invocation and must honor context cancellation.
type Probe func(context.Context) error

// Component is one independently managed part of a Service.
//
// Start performs bounded initialization and is called in registration order.
// Run starts only after every component has started successfully. The first
// Run function to return initiates shutdown; a nil return is graceful and a
// non-nil return fails the service. Drain is called in reverse registration
// order before Run contexts are canceled, allowing listeners and claim loops to
// stop admitting new work. Stop is then called in reverse registration order.
// Components without Run are passive lifecycle participants.
type Component struct {
	Name      string
	Start     Hook
	Run       Hook
	Drain     Hook
	Stop      Hook
	Liveness  Probe
	Readiness Probe
}

func (component Component) validate() error {
	if err := validateName("component", component.Name, operationAdd); err != nil {
		return err
	}
	if component.Start == nil && component.Run == nil && component.Drain == nil && component.Stop == nil &&
		component.Liveness == nil && component.Readiness == nil {
		return nilComponentError(component.Name)
	}
	return nil
}

func validateName(kind, name, operation string) error {
	if name == "" || len(name) > maxNameLength || strings.TrimSpace(name) != name {
		return invalidNameError(kind, name, operation)
	}

	for index, r := range name {
		if unicode.IsLetter(r) || unicode.IsDigit(r) {
			continue
		}
		if index > 0 && (r == '-' || r == '_' || r == '.' || r == '/') {
			continue
		}
		return invalidNameError(kind, name, operation)
	}
	return nil
}

func invoke(ctx context.Context, function func(context.Context) error) (err error) {
	if function == nil {
		return nil
	}
	defer func() {
		if recovered := recover(); recovered != nil {
			err = &PanicError{Value: recovered, stack: debug.Stack()}
		}
	}()
	return function(ctx)
}

// invokeBounded prevents the lifecycle coordinator from blocking forever when
// a hook ignores cancellation. A hook that violates the contract may continue
// in its goroutine after this function returns and can leak work.
func invokeBounded(ctx context.Context, function func(context.Context) error) error {
	if function == nil {
		return nil
	}

	result := make(chan error, 1)
	go func() {
		result <- invoke(ctx, function)
	}()

	select {
	case err := <-result:
		return err
	case <-ctx.Done():
		return ctx.Err()
	}
}

File: libs/rust/servicekit/src/component.rs
// Copyright © 2026 Mindclade, LLC. All Rights Reserved.
// Mindclade Proprietary and Confidential.
// SPDX-License-Identifier: LicenseRef-Mindclade-Proprietary
//

//! Components registered with the deterministic service lifecycle.

use mindclade_faults::FaultResult;

/// A process-owned component with explicit startup, drain, and stop hooks.
///
/// Implementations must make `drain` and `stop` idempotent. `Service` may call
/// `stop` during startup rollback after a later component fails to start.
pub trait Component: Send {
    /// Stable component name used in diagnostics and failure context.
    fn name(&self) -> &str;
    /// Acquire resources and start background work.
    fn start(&mut self) -> FaultResult<()>;
    /// Reject new work while allowing already-admitted work to finish.
    fn drain(&mut self) -> FaultResult<()> {
        Ok(())
    }
    /// Release resources and terminate background work.
    fn stop(&mut self) -> FaultResult<()>;
}

/// Adapter for small components whose lifecycle is represented by closures.
pub struct FnComponent<S, D, T>
where
    S: FnMut() -> FaultResult<()> + Send,
    D: FnMut() -> FaultResult<()> + Send,
    T: FnMut() -> FaultResult<()> + Send,
{
    name: String,
    start: S,
    drain: D,
    stop: T,
}

impl<S, D, T> FnComponent<S, D, T>
where
    S: FnMut() -> FaultResult<()> + Send,
    D: FnMut() -> FaultResult<()> + Send,
    T: FnMut() -> FaultResult<()> + Send,
{
    #[must_use]
    pub fn new(name: impl Into<String>, start: S, drain: D, stop: T) -> Self {
        Self {
            name: name.into(),
            start,
            drain,
            stop,
        }
    }
}

impl<S, D, T> Component for FnComponent<S, D, T>
where
    S: FnMut() -> FaultResult<()> + Send,
    D: FnMut() -> FaultResult<()> + Send,
    T: FnMut() -> FaultResult<()> + Send,
{
    fn name(&self) -> &str {
        &self.name
    }
    fn start(&mut self) -> FaultResult<()> {
        (self.start)()
    }
    fn drain(&mut self) -> FaultResult<()> {
        (self.drain)()
    }
    fn stop(&mut self) -> FaultResult<()> {
        (self.stop)()
    }
}
